packages/react/toggle-button/src/ToggleButton.tsx

```diff
@@ -1,10 +1,199 @@
+import { RovingFocusGroup, useRovingFocus } from '@radix-ui/react-roving-focus';
 import { forwardRefWithAs } from '@radix-ui/react-polymorphic';
 
-const NAME = 'ToggleButton';
-const DEFAULT_TAG = 'button';
+const __DEV__ = isDev();
+
+// Config 1: Multiple selection, at least one required
+// Config 2: Multiple selection, none required
+// Config 3: Single selection,  at least one required
+// Config 4: Single selection, none required
+
+/* -------------------------------------------------------------------------------------------------
+ * ToggleButtonGroup
+ * -----------------------------------------------------------------------------------------------*/
+
+const GROUP_NAME = 'ToggleButtonGroup';
+const GROUP_DEFAULT_TAG = 'div';
+const GROUP_CONTEXT_NAME = 'ToggleButtonGroupContext';
+
+type ToggleButtonGroupContextValue = {
+  value: string[];
+  setValue: React.Dispatch<React.SetStateAction<string[]>>;
+  handleChange(value: string): void;
+  name: typeof GROUP_CONTEXT_NAME;
+  rovingFocus: boolean;
+};
+
+const ToggleButtonGroupContext = React.createContext<ToggleButtonGroupContextValue | null>(null);
+ToggleButtonGroupContext.displayName = GROUP_CONTEXT_NAME;
+
+type ToggleButtonGroupSharedProps = {
+  /** Whether or not a selection in the group is required after initial selection */
+  selectionIsRequired?: boolean;
+  /** Whether or not the group should maintain roving focus of its buttons */
+  rovingFocus?: boolean;
+};
+
+type ToggleButtonGroupOwnProps = ToggleButtonGroupSharedProps & {
+  /** The controlled value of the toggled button */
+  value?: string[];
+  /** The uncontrolled value of the toggled button */
+  defaultValue?: string[];
+  /** A function called when the value of the toggled buttons changes */
+  onValueChange?(value: string[]): void;
+};
+
+const ToggleButtonGroup = forwardRefWithAs<typeof GROUP_DEFAULT_TAG, ToggleButtonGroupOwnProps>(
+  (props, forwardedRef) => {
+    const {
+      as: Comp = GROUP_DEFAULT_TAG,
+      value: valueProp,
+      defaultValue,
+      onValueChange,
+      children,
+      selectionIsRequired = false,
+      rovingFocus = false,
+      ...groupProps
+    } = props;
+
+    const [value, setValue] = useControlledState<string[]>({
+      prop: valueProp,
+      defaultProp: defaultValue || [],
+      onChange: onValueChange,
+      isEqual: stringArraysAreEqual,
+    });
+
+    function handleChange(buttonValue: string) {
+      if (!value || value.length < 1) {
+        return setValue([buttonValue]);
+      }
+
+      if (selectionIsRequired && value.length === 1 && value.includes(buttonValue)) {
+        return;
+      }
+
+      setValue(
+        value.includes(buttonValue)
+          ? value.filter((v) => v !== buttonValue)
+          : value.concat(buttonValue).sort()
+      );
+    }
+
+    const context: ToggleButtonGroupContextValue = {
+      setValue,
+      value,
+      handleChange,
+      rovingFocus,
+      name: GROUP_CONTEXT_NAME,
+    };
```

Comment:
Oh yeah, I forgot to leave a note on all the `context` declarations in this file too... we need to `React.useMemo` these to avoid re-rendering consumers all the time.

Reply:
Typically I would have, but in this case any state/rendering change is almost certainly going to be because the toggle state changes anyway, meaning we are probably memoing for naught. Fine to do it if that's just something we want to do as default but I suspect it won't make a huge difference here.

Reply:
I'd prefer if we could be more explicit personally because what you described is some implied knowledge of the inner workings. Sure, it might re-memoize every-time but as another developer reading your code, I would be able to see why 🙏